awx/main/fields.py

```diff
@@ -858,27 +856,41 @@ def __str__(self):
                 template_name = template_name.split('.')[1]
                 setattr(valid_namespace['tower'].filename, template_name, 'EXAMPLE_FILENAME')
 
+        def validate_template_string(type_, key, tmpl):
+            try:
+                sandbox.ImmutableSandboxedEnvironment(undefined=StrictUndefined).from_string(tmpl).render(valid_namespace)
+            except UndefinedError as e:
+                raise django_exceptions.ValidationError(
+                    _('{sub_key} uses an undefined field ({error_msg})').format(sub_key=key, error_msg=e),
+                    code='invalid',
+                    params={'value': value},
+                )
+            except SecurityError as e:
+                raise django_exceptions.ValidationError(_('Encountered unsafe code execution: {}').format(e))
+            except TemplateSyntaxError as e:
+                raise django_exceptions.ValidationError(
+                    _('Syntax error rendering template for {sub_key} inside of {type} ({error_msg})').format(sub_key=key, type=type_, error_msg=e),
+                    code='invalid',
+                    params={'value': value},
+                )
+
+        def validate_extra_vars(key, node):
+            if isinstance(node, dict):
+                return {validate_extra_vars(key, k): validate_extra_vars(k, v) for k, v in node.items()}
```

Comment:
We really don't want to lose context in error messages. The `key` passed in should always be a string right? So what if you combined `k` and `key` for the value here, so instead of `validate_extra_vars(k, v)` do `validate_extra_vars(f'{key}:{k}', v)`, or whatever you'd prefer for delimiter. That would keep stacking layers of context on a rolling basis.

Reply:
I used a '.' (dot) as the dict key delimiter, seemed appropriate for this.